pkg/controller/provider/container/vsphere/model.go

```diff
@@ -790,6 +797,20 @@ func (v *VmAdapter) Apply(u types.ObjectUpdate) {
 			}
 		}
 	}
+
+	// There might be cases which GuestNetworks.origin (Which indicates the ip address type )is missing
+	// So we need to mark (if possible) the static ip address by setting the GuestNetworks.origin field to manual.
+	if !skipStaticIpAssign {
+		for _, nic := range v.model.NICs {
+			for i := range v.model.GuestNetworks {
+				guestNetwork := &v.model.GuestNetworks[i]
+				if strings.TrimSpace(guestNetwork.Origin) == "" && nic.MAC == guestNetwork.MAC {
```

Comment:
if we are testing for the existence of a nic with specific MAC, we can replace the nested loop with a dictionary.

``` go
	// Build a map of NICs by lowercase MAC address
	nicMap := make(map[string]struct{})
	for _, nic := range v.model.NICs {
		nicMap[strings.ToLower(nic.MAC)] = struct{}{}
	}

	// Iterate over guest networks and assign origin if MAC is in NICs
	for i := range v.model.GuestNetworks {
		guestNetwork := &v.model.GuestNetworks[i]
		if strings.TrimSpace(guestNetwork.Origin) == "" {
			if _, found := nicMap[strings.ToLower(guestNetwork.MAC)]; found {
				guestNetwork.Origin = string(types.NetIpConfigInfoIpAddressOriginManual)
			}
		}
	}
	```

Reply:
p.s.
if you check that you don't overwrite the origin if exist, you don't need the `skipStaticIpAssign` check

Reply:
p.p.s
can be refactored into a method with meaningful name

Reply:
p.p.p.s
this is a hack we should only use on none windows machine, on windows we can trust the value of the `origin`  we only need to guess the origin using this hack if we have a linux machine.

a. we should add a guard to use this only on none windows machines
b. we should add a comment that this is a hack